# Introduction to Pandas

Think of Pandas as Python version of Excel.

# Series

The first main data type we will learn about for pandas is the Series data type. 

A Series is very similar to a NumPy array (in fact it is built on top of the NumPy array object). 

What differentiates the NumPy array from a Series, is that a Series can have axis labels, meaning it can be indexed by a label, instead of just a number location. 

It also doesn't need to hold numeric data, it can hold any arbitrary Python Object.

In [1]:
import numpy as np
import pandas as pd

## Creating a Series
You can convert a list, numpy array, or dictionary to a Series.

In [2]:
labels = ['a', 'b', 'c']
my_list = [10, 20, 30]
arr = np.array([10, 20, 30])
d = {"a": 10, "b": 20, "c": 30}

### List

In [3]:
pd.Series(data=my_list)

0    10
1    20
2    30
dtype: int64

In [4]:
pd.Series(data=my_list, index=labels)

a    10
b    20
c    30
dtype: int64

In [5]:
pd.Series(my_list, labels)

a    10
b    20
c    30
dtype: int64

### NumPy Array

In [6]:
pd.Series(arr)

0    10
1    20
2    30
dtype: int64

In [7]:
pd.Series(arr, labels)

a    10
b    20
c    30
dtype: int64

### Dictionary

In [8]:
pd.Series(d)

a    10
b    20
c    30
dtype: int64

## Data in Series

A Pandas Series can hold a variety of object types.

In [9]:
pd.Series(data=labels)

0    a
1    b
2    c
dtype: object

In [10]:
# Even functions (although unlikely that you will use this)
pd.Series([sum, print, len])

0      <built-in function sum>
1    <built-in function print>
2      <built-in function len>
dtype: object

## Using an Index
The key to using a Series is understanding its index. 

Pandas makes use of these index names or numbers by allowing for fast look ups of information (works like a hash table or dictionary).

In [10]:
ser1 = pd.Series([1, 2, 3, 4], index=["USA", "Germany", "USSR", "Japan"])

In [11]:
ser1

USA        1
Germany    2
USSR       3
Japan      4
dtype: int64

In [12]:
ser2 = pd.Series([1, 2, 5, 4], index = ["USA", "Germany", "Italy", "Japan"])

In [13]:
ser2

USA        1
Germany    2
Italy      5
Japan      4
dtype: int64

In [14]:
ser1["USA"] # we use index label here

1

Operations are then also done based off of index.

In [16]:
ser1 + ser2

Germany    4.0
Italy      NaN
Japan      8.0
USA        2.0
USSR       NaN
dtype: float64

# DataFrames

DataFrames are the workhorse of pandas and are directly inspired by the R programming language. We can think of a DataFrame as a bunch of Series objects put together to share the same index.

In [16]:
import pandas as pd
import numpy as np

In [17]:
# Create DataFrame with random data.
np.random.seed(101)
df = pd.DataFrame(np.random.randn(5, 4), index=["A", "B", "C", "D", "E"], columns=["W", "X", "Y", "Z"])
df


W         X         Y         Z
A  2.706850  0.628133  0.907969  0.503826
B  0.651118 -0.319318 -0.848077  0.605965
C -2.018168  0.740122  0.528813 -0.589001
D  0.188695 -0.758872 -0.933237  0.955057
E  0.190794  1.978757  2.605967  0.683509

Each column here is a Series. That what Pandas DataFrame is: a bunch of Series that share common index.

## Selection and Indexing

### Select by column name.

In [18]:
df['W']

A    2.706850
B    0.651118
C   -2.018168
D    0.188695
E    0.190794
Name: W, dtype: float64

The result looks like a Series.

In [19]:
type(df["W"])

pandas.core.series.Series

If you select multiple columns - use list - you get a DataFrame.

In [6]:
df[['W', 'Z']]

W         Z
A  2.706850  0.503826
B  0.651118  0.605965
C -2.018168 -0.589001
D  0.188695  0.955057
E  0.190794  0.683509

In [8]:
type(df[['W', 'Z']])

pandas.core.frame.DataFrame

### SQL Syntax (NOR RECOMMENDED!)

In [21]:
df.W

A    2.706850
B    0.651118
C   -2.018168
D    0.188695
E    0.190794
Name: W, dtype: float64

In [10]:
df.#click tab now to see all DataFrames methods!

SyntaxError: invalid syntax (<ipython-input-10-dfdb3d2d8502>, line 1)

### Creating a new column

Use dictionary like syntax.

In [23]:
df['new'] = df['W'] + df['Y']

In [24]:
df

W         X         Y         Z       new
A  2.706850  0.628133  0.907969  0.503826  3.614819
B  0.651118 -0.319318 -0.848077  0.605965 -0.196959
C -2.018168  0.740122  0.528813 -0.589001 -1.489355
D  0.188695 -0.758872 -0.933237  0.955057 -0.744542
E  0.190794  1.978757  2.605967  0.683509  2.796762

### Removing columns

drop method by default drops axis=0 which means it drops index. So to drop columns we need to specify axis=1.

In [22]:
df.drop("Z", axis=1)

W         X         Y
A  2.706850  0.628133  0.907969
B  0.651118 -0.319318 -0.848077
C -2.018168  0.740122  0.528813
D  0.188695 -0.758872 -0.933237
E  0.190794  1.978757  2.605967

In [23]:
df.drop("A", axis=0)

W         X         Y         Z
B  0.651118 -0.319318 -0.848077  0.605965
C -2.018168  0.740122  0.528813 -0.589001
D  0.188695 -0.758872 -0.933237  0.955057
E  0.190794  1.978757  2.605967  0.683509

It is not in-place unless specified!

In [26]:
df

W         X         Y         Z       new
A  2.706850  0.628133  0.907969  0.503826  3.614819
B  0.651118 -0.319318 -0.848077  0.605965 -0.196959
C -2.018168  0.740122  0.528813 -0.589001 -1.489355
D  0.188695 -0.758872 -0.933237  0.955057 -0.744542
E  0.190794  1.978757  2.605967  0.683509  2.796762

In [27]:
df.drop("new", axis=1, inplace=True) 

In [28]:
df

W         X         Y         Z
A  2.706850  0.628133  0.907969  0.503826
B  0.651118 -0.319318 -0.848077  0.605965
C -2.018168  0.740122  0.528813 -0.589001
D  0.188695 -0.758872 -0.933237  0.955057
E  0.190794  1.978757  2.605967  0.683509

### Removing rows

In [13]:
df.drop("E", axis=0) # not in-place!

W         X         Y         Z
A  2.706850  0.628133  0.907969  0.503826
B  0.651118 -0.319318 -0.848077  0.605965
C -2.018168  0.740122  0.528813 -0.589001
D  0.188695 -0.758872 -0.933237  0.955057

### Selecting rows

Selecting columns is above.

### Select based on index label.

In [14]:
df

W         X         Y         Z
A  2.706850  0.628133  0.907969  0.503826
B  0.651118 -0.319318 -0.848077  0.605965
C -2.018168  0.740122  0.528813 -0.589001
D  0.188695 -0.758872 -0.933237  0.955057
E  0.190794  1.978757  2.605967  0.683509

We use loc method which stands for location.

We use square brackets!

In [30]:
df.loc['A']

W    2.706850
X    0.628133
Y    0.907969
Z    0.503826
Name: A, dtype: float64

It returns row Series.

### Select based on index position instead of label.

Grab by the row index number.

In [31]:
df.iloc[0]

W    2.706850
X    0.628133
Y    0.907969
Z    0.503826
Name: A, dtype: float64

### Selecting subset of rows and columns

We simply specify row and column like in NumPy. 

We use loc methods here!

In [32]:
df.loc['B', 'Y'] # row, column 

-0.84807698340363147

In [33]:
df.loc[["A", "B"], ["W", "Y"]]

W         Y
A  2.706850  0.907969
B  0.651118 -0.848077

## Conditional selection

An important feature of Pandas is conditional selection using bracket notation, very similar to NumPy.

### Entire DataFrame.

In [34]:
df

W         X         Y         Z
A  2.706850  0.628133  0.907969  0.503826
B  0.651118 -0.319318 -0.848077  0.605965
C -2.018168  0.740122  0.528813 -0.589001
D  0.188695 -0.758872 -0.933237  0.955057
E  0.190794  1.978757  2.605967  0.683509

In [35]:
df > 0

W      X      Y      Z
A   True   True   True   True
B   True  False  False   True
C  False   True   True  False
D   True  False  False   True
E   True   True   True   True

In [36]:
df[df > 0]

W         X         Y         Z
A  2.706850  0.628133  0.907969  0.503826
B  0.651118       NaN       NaN  0.605965
C       NaN  0.740122  0.528813       NaN
D  0.188695       NaN       NaN  0.955057
E  0.190794  1.978757  2.605967  0.683509

### Conditional selection based on specific rows or columns.

We create boolean Series and then use it to select values.

In [15]:
df["W"] > 0

A     True
B     True
C    False
D     True
E     True
Name: W, dtype: bool

In [37]:
df[df["W"] > 0]

W         X         Y         Z
A  2.706850  0.628133  0.907969  0.503826
B  0.651118 -0.319318 -0.848077  0.605965
D  0.188695 -0.758872 -0.933237  0.955057
E  0.190794  1.978757  2.605967  0.683509

Selecting by rows or columns results with DataFrame with only rows and columns passing the condition - no Nan values. 

It is the most often used version of conditional selection.

In this example we will get only one row passing condition and then select column from the result.

In [18]:
df[df['W'] < 0]

W         X         Y         Z
C -2.018168  0.740122  0.528813 -0.589001

In [19]:
df[df['W'] < 0]["Y"]

C    0.528813
Name: Y, dtype: float64

In [39]:
df[df["W"] > 0][["Y", "X"]]

Y         X
A  0.907969  0.628133
B -0.848077 -0.319318
D -0.933237 -0.758872
E  2.605967  1.978757

For two conditions you can use | and & with parenthesis.

Not and/or - this will cause error.

```
The truth value of a Series is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().
```
A Python and operator can't copare a series to a series. It takes single booleans at a time. When passing entire Series it gets confused.

Use `&` and `|`. And remember to use brackets for conditions.

In [20]:
df[(df["W"] > 0) and (df["Y"] > 1)]

ValueError: The truth value of a Series is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().

In [21]:
df[(df["W"] > 0) & (df["Y"] > 1)]

W         X         Y         Z
E  0.190794  1.978757  2.605967  0.683509

In [22]:
df[(df["W"] > 0) | (df["Y"] > 1)]

W         X         Y         Z
A  2.706850  0.628133  0.907969  0.503826
B  0.651118 -0.319318 -0.848077  0.605965
D  0.188695 -0.758872 -0.933237  0.955057
E  0.190794  1.978757  2.605967  0.683509

## More Index details

Let's discuss some more features of indexing, including resetting the index or setting it something else. 

We'll also talk about index hierarchy

In [24]:
df

index         W         X         Y         Z
0     A  2.706850  0.628133  0.907969  0.503826
1     B  0.651118 -0.319318 -0.848077  0.605965
2     C -2.018168  0.740122  0.528813 -0.589001
3     D  0.188695 -0.758872 -0.933237  0.955057
4     E  0.190794  1.978757  2.605967  0.683509

### Reset to default 0,1...n index.

In [42]:
df.reset_index()

index         W         X         Y         Z
0     A  2.706850  0.628133  0.907969  0.503826
1     B  0.651118 -0.319318 -0.848077  0.605965
2     C -2.018168  0.740122  0.528813 -0.589001
3     D  0.188695 -0.758872 -0.933237  0.955057
4     E  0.190794  1.978757  2.605967  0.683509

We got index reset to a column ans actual index to be numerical.

Also does not occur in-place unless specified.

### Set index manually.

Add new column and the set that column an index.

In [43]:
newind = ["CA", "NY", "WY", "OR", "CO"]

In [44]:
df['States'] = newind

In [45]:
df

W         X         Y         Z States
A  2.706850  0.628133  0.907969  0.503826     CA
B  0.651118 -0.319318 -0.848077  0.605965     NY
C -2.018168  0.740122  0.528813 -0.589001     WY
D  0.188695 -0.758872 -0.933237  0.955057     OR
E  0.190794  1.978757  2.605967  0.683509     CO

In [46]:
df.set_index("States")

W         X         Y         Z
States                                        
CA      2.706850  0.628133  0.907969  0.503826
NY      0.651118 -0.319318 -0.848077  0.605965
WY     -2.018168  0.740122  0.528813 -0.589001
OR      0.188695 -0.758872 -0.933237  0.955057
CO      0.190794  1.978757  2.605967  0.683509

Again happened in-place.

In [47]:
df

W         X         Y         Z States
A  2.706850  0.628133  0.907969  0.503826     CA
B  0.651118 -0.319318 -0.848077  0.605965     NY
C -2.018168  0.740122  0.528813 -0.589001     WY
D  0.188695 -0.758872 -0.933237  0.955057     OR
E  0.190794  1.978757  2.605967  0.683509     CO

In [48]:
df.set_index("States", inplace=True)

In [49]:
df

W         X         Y         Z
States                                        
CA      2.706850  0.628133  0.907969  0.503826
NY      0.651118 -0.319318 -0.848077  0.605965
WY     -2.018168  0.740122  0.528813 -0.589001
OR      0.188695 -0.758872 -0.933237  0.955057
CO      0.190794  1.978757  2.605967  0.683509

## Multi-Index and Index hierarchy

Let us go over how to work with Multi-Index.

First we'll create a quick example of what a Multi-Indexed DataFrame would look like.

In [24]:
# Index Levels
outside = ["G1", "G1", "G1", "G2", "G2", "G2"]
inside = [1, 2, 3, 1, 2, 3]
hier_index = list(zip(outside, inside))
hier_index

[('G1', 1), ('G1', 2), ('G1', 3), ('G2', 1), ('G2', 2), ('G2', 3)]

In [27]:
hier_index = pd.MultiIndex.from_tuples(hier_index)
hier_index

MultiIndex(levels=[['G1', 'G2'], [1, 2, 3]],
           labels=[[0, 0, 0, 1, 1, 1], [0, 1, 2, 0, 1, 2]])

In [28]:
df = pd.DataFrame(np.random.randn(6, 2), index=hier_index, columns=['A', 'B'])
df

A         B
G1 1  0.302665  1.693723
   2 -1.706086 -1.159119
   3 -0.134841  0.390528
G2 1  0.166905  0.184502
   2  0.807706  0.072960
   3  0.638787  0.329646

Now let's show how to index this! 

For index hierarchy we use df.loc[], if this was on the columns axis, you would just use normal bracket notation df[]. 

Calling one level of the index returns the sub-dataframe.

In [53]:
df.loc["G1"]




A         B
1  0.302665  1.693723
2 -1.706086 -1.159119
3 -0.134841  0.390528

Calling two levels of the index returns the Series.

In [54]:
df.loc["G1"].loc[1]

A    0.302665
B    1.693723
Name: 1, dtype: float64

You can name you inexes.

In [37]:
df.index.names

FrozenList([None, None])

In [38]:
df.index.names = ["Group", "Num"]

In [39]:
df

A         B
Group Num                    
G1    1    0.302665  1.693723
      2   -1.706086 -1.159119
      3   -0.134841  0.390528
G2    1    0.166905  0.184502
      2    0.807706  0.072960
      3    0.638787  0.329646

Imagine you want to grab value "0.072960".

It is G2, 2, column B.

In [31]:
df.loc['G2'].loc[2]['B']

0.072959675317038689

Cross function returns a cross-section (row(s) or column(s)) from the
Series/DataFrame.

In [32]:
df.xs("G1")

A         B
1  0.302665  1.693723
2 -1.706086 -1.159119
3 -0.134841  0.390528

I want to select rows with Num=1 from both groups.

In [40]:
df.xs(1, level="Num")

A         B
Group                    
G1     0.302665  1.693723
G2     0.166905  0.184502

In [59]:
df.xs(["G1", 1])

A    0.302665
B    1.693723
Name: (G1, 1), dtype: float64

# Missing data

Let's show a few convenient methods to deal with Missing Data in Pandas.

In [41]:
df = pd.DataFrame({"A": [1, 2, np.nan],
                   "B": [5, np.nan, np.nan],
                   "C": [1, 2, 3]})

In [42]:
df

A    B  C
0  1.0  5.0  1
1  2.0  NaN  2
2  NaN  NaN  3

Throw away rows with missing values in it.

In [43]:
df.dropna()

A    B  C
0  1.0  5.0  1

HINT:
In Pandas axis=0 means that operation occurs along the rows.

axis=1 means that we do things along the columns.

Check it against NumPy.

If we want to drop columns with NaNs use axis=1.

In [44]:
df.dropna(axis=1)

C
0  1
1  2
2  3

We can specify threshold for number of NaNs for row or column to get dropped.

In [65]:
df.dropna(thresh=2)

A    B  C
0  1.0  5.0  1
1  2.0  NaN  2

## Fillna

We do not want to drop the values - just change the values.

In [66]:
df.fillna(value="Fill VALUE")

A           B  C
0           1           5  1
1           2  Fill VALUE  2
2  Fill VALUE  Fill VALUE  3

Fill missing value with the mean of that column.

In [67]:
df["A"].fillna(value=df["A"].mean())

0    1.0
1    2.0
2    1.5
Name: A, dtype: float64

It is really important to think how you fill missing values.

# Groupby

The groupby method allows you to group rows of data together based off of a column and perform an aggregate function on them.

In [46]:
data = {"Company": ["GOOG", "GOOG", "MSFT", "MSFT", "FB", "FB"],
        "Person": ["Sam", "Charlie", "Amy", "Vanessa", "Carl", "Sarah"],
        "Sales": [200, 120, 340, 124, 243, 350]}
df = pd.DataFrame(data)
df

Company   Person  Sales
0    GOOG      Sam    200
1    GOOG  Charlie    120
2    MSFT      Amy    340
3    MSFT  Vanessa    124
4      FB     Carl    243
5      FB    Sarah    350

Now you can use the .groupby() method to group rows together based off of a column name. 

For instance let's group based off of Company. This will create a DataFrameGroupBy object.

In [69]:
df.groupby("Company")

You can save this object as a new variable.

In [48]:
by_comp = df.groupby("Company")

And then call aggregate methods off the object.

HINT: Potrzebuję na dłużej popracować z Pandasem bez wizualizacji - nabrać biegłości we wszystkich tych operacjach.

In [49]:
by_comp.mean()

Sales
Company       
FB       296.5
GOOG     160.0
MSFT     232.0

In [50]:
by_comp.sum()

Sales
Company       
FB         593
GOOG       320
MSFT       464

In [52]:
by_comp.sum().loc["FB"]

Sales    593
Name: FB, dtype: int64

In [54]:
df.groupby("Company").mean().loc["FB"]

Sales    296.5
Name: FB, dtype: float64

More examples of aggregate methods.

In [73]:
by_comp.std()

Sales
Company            
FB        75.660426
GOOG      56.568542
MSFT     152.735065

In [74]:
by_comp.min()

Person  Sales
Company                
FB          Carl    243
GOOG     Charlie    120
MSFT         Amy    124

In [75]:
by_comp.max()

Person  Sales
Company                
FB         Sarah    350
GOOG         Sam    200
MSFT     Vanessa    340

In [76]:
by_comp.count()

Person  Sales
Company               
FB            2      2
GOOG          2      2
MSFT          2      2

In [77]:
by_comp.describe()

Sales
Company                  
FB      count    2.000000
        mean   296.500000
        std     75.660426
        min    243.000000
        25%    269.750000
        50%    296.500000
        75%    323.250000
        max    350.000000
GOOG    count    2.000000
        mean   160.000000
        std     56.568542
        min    120.000000
        25%    140.000000
        50%    160.000000
        75%    180.000000
        max    200.000000
MSFT    count    2.000000
        mean   232.000000
        std    152.735065
        min    124.000000
        25%    178.000000
        50%    232.000000
        75%    286.000000
        max    340.000000

In [78]:
by_comp.describe().transpose()

Company    FB                                                         GOOG  \
        count   mean        std    min     25%    50%     75%    max count   
Sales     2.0  296.5  75.660426  243.0  269.75  296.5  323.25  350.0   2.0   

Company         ...                  MSFT                                   \
          mean  ...      75%    max count   mean         std    min    25%   
Sales    160.0  ...    180.0  200.0   2.0  232.0  152.735065  124.0  178.0   

Company                       
           50%    75%    max  
Sales    232.0  286.0  340.0  

[1 rows x 24 columns]

In [79]:
by_comp.describe().transpose()['GOOG']

count   mean        std    min    25%    50%    75%    max
Sales    2.0  160.0  56.568542  120.0  140.0  160.0  180.0  200.0

# Merging, Joining, and Concatenating

There are 3 main ways of combining DataFrames together: Merging, Joining and Concatenating.

## Concatenation

Concatenation basically glues together DataFrames. Keep in mind that dimensions should match along the axis you are concatenating on. You can use **pd.concat** and pass in a list of DataFrames to concatenate together.

In [25]:
df1 = pd.DataFrame({"A": ["A0", "A1", "A2", "A3"],
                    "B": ["B0", "B1", "B2", "B3"],
                    "C": ["C0", "C1", "C2", "C3"],
                    "D": ["D0", "D1", "D2", "D3"]},
                  index=[0, 1, 2, 3])

In [26]:
df2 = pd.DataFrame({"A": ["A4", "A5", "A6", "A7"],
                    "B": ["B4", "B5", "B6", "B7"],
                    "C": ["C4", "C5", "C6", "C7"],
                    "D": ["D4", "D5", "D6", "D7"]},
                  index=[4, 5, 6, 7])

In [28]:
df3 = pd.DataFrame({"A": ["A8", "A9", "A10", "A11"],
                    "B": ["B8", "B9", "B10", "B11"],
                    "C": ["C8", "C9", "C10", "C11"],
                    "D": ["D8", "D9", "D10", "D11"]},
                  index=[8, 9, 10, 11])

In [83]:
df1

A   B   C   D
0  A0  B0  C0  D0
1  A1  B1  C1  D1
2  A2  B2  C2  D2
3  A3  B3  C3  D3

In [84]:
df2

A   B   C   D
4  A4  B4  C4  D4
5  A5  B5  C5  D5
6  A6  B6  C6  D6
7  A7  B7  C7  D7

In [85]:
df3

A    B    C    D
8    A8   B8   C8   D8
9    A9   B9   C9   D9
10  A10  B10  C10  D10
11  A11  B11  C11  D11

In [29]:
pd.concat([df1, df2, df3])

A    B    C    D
0    A0   B0   C0   D0
1    A1   B1   C1   D1
2    A2   B2   C2   D2
3    A3   B3   C3   D3
4    A4   B4   C4   D4
5    A5   B5   C5   D5
6    A6   B6   C6   D6
7    A7   B7   C7   D7
8    A8   B8   C8   D8
9    A9   B9   C9   D9
10  A10  B10  C10  D10
11  A11  B11  C11  D11

Concatenate along columns.

In [30]:
pd.concat([df1, df2, df3], axis=1)

A    B    C    D    A    B    C    D
0   A0   B0   C0   D0  NaN  NaN  NaN  NaN
1   A1   B1   C1   D1  NaN  NaN  NaN  NaN
2   A2   B2   C2   D2  NaN  NaN  NaN  NaN
3   A3   B3   C3   D3  NaN  NaN  NaN  NaN
4  NaN  NaN  NaN  NaN   A4   B4   C4   D4
5  NaN  NaN  NaN  NaN   A5   B5   C5   D5
6  NaN  NaN  NaN  NaN   A6   B6   C6   D6
7  NaN  NaN  NaN  NaN   A7   B7   C7   D7

## Merging

The **merge** function allows you to merge DataFrames together using a similar logic as merging SQL Tables together.

In [88]:
left = pd.DataFrame({"key": ["K0", "K1", "K2", "K3"],
                     "A": ["A0", "A1", "A2", "A3"],
                     "B": ["B0", "B1", "B2", "B3"]})

right = pd.DataFrame({"key": ["K0", "K1", "K2", "K3"],
                      "C": ["C0", "C1", "C2", "C3"],
                      "D": ["D0", "D1", "D2", "D3"]})

In [89]:
left

A   B key
0  A0  B0  K0
1  A1  B1  K1
2  A2  B2  K2
3  A3  B3  K3

In [90]:
right

C   D key
0  C0  D0  K0
1  C1  D1  K1
2  C2  D2  K2
3  C3  D3  K3

In [91]:
pd.merge(left, right, how="inner", on="key")

A   B key   C   D
0  A0  B0  K0  C0  D0
1  A1  B1  K1  C1  D1
2  A2  B2  K2  C2  D2
3  A3  B3  K3  C3  D3

Or to show a more complicated example.

In [32]:
left = pd.DataFrame({'key1': ['K0', 'K0', 'K1', 'K2'],
                     'key2': ['K0', 'K1', 'K0', 'K1'],
                        'A': ['A0', 'A1', 'A2', 'A3'],
                        'B': ['B0', 'B1', 'B2', 'B3']})
    
right = pd.DataFrame({'key1': ['K0', 'K1', 'K1', 'K2'],
                               'key2': ['K0', 'K0', 'K0', 'K0'],
                                  'C': ['C0', 'C1', 'C2', 'C3'],
                                  'D': ['D0', 'D1', 'D2', 'D3']})

In [33]:
left

A   B key1 key2
0  A0  B0   K0   K0
1  A1  B1   K0   K1
2  A2  B2   K1   K0
3  A3  B3   K2   K1

In [34]:
right

C   D key1 key2
0  C0  D0   K0   K0
1  C1  D1   K1   K0
2  C2  D2   K1   K0
3  C3  D3   K2   K0

In [37]:
pd.merge(left, right, how="inner", on=["key1", "key2"])

A   B key1 key2   C   D
0  A0  B0   K0   K0  C0  D0
1  A2  B2   K1   K0  C1  D1
2  A2  B2   K1   K0  C2  D2

In [96]:
pd.merge(left, right, how="outer", on=["key1", "key2"])

A    B key1 key2    C    D
0   A0   B0   K0   K0   C0   D0
1   A1   B1   K0   K1  NaN  NaN
2   A2   B2   K1   K0   C1   D1
3   A2   B2   K1   K0   C2   D2
4   A3   B3   K2   K1  NaN  NaN
5  NaN  NaN   K2   K0   C3   D3

In [97]:
pd.merge(left, right, how="right", on=["key1", "key2"])

A    B key1 key2   C   D
0   A0   B0   K0   K0  C0  D0
1   A2   B2   K1   K0  C1  D1
2   A2   B2   K1   K0  C2  D2
3  NaN  NaN   K2   K0  C3  D3

In [98]:
pd.merge(left, right, how="left", on=["key1", "key2"])

A   B key1 key2    C    D
0  A0  B0   K0   K0   C0   D0
1  A1  B1   K0   K1  NaN  NaN
2  A2  B2   K1   K0   C1   D1
3  A2  B2   K1   K0   C2   D2
4  A3  B3   K2   K1  NaN  NaN

## Joining

Joining is a convenient method for combining the columns of two potentially differently-indexed DataFrames into a single result DataFrame.

You can think about it as the same thing as merge except the keys you want to join on are on you index instead of a column.

In [38]:
left = pd.DataFrame({'A': ['A0', 'A1', 'A2'],
                     'B': ['B0', 'B1', 'B2']},
                      index=['K0', 'K1', 'K2']) 

right = pd.DataFrame({'C': ['C0', 'C2', 'C3'],
                    'D': ['D0', 'D2', 'D3']},
                      index=['K0', 'K2', 'K3'])

In [100]:
left

A   B
K0  A0  B0
K1  A1  B1
K2  A2  B2

In [101]:
right

C   D
K0  C0  D0
K2  C2  D2
K3  C3  D3

In [40]:
left.join(right, how="left")

A   B    C    D
K0  A0  B0   C0   D0
K1  A1  B1  NaN  NaN
K2  A2  B2   C2   D2

In [103]:
left.join(right, how="outer")

A    B    C    D
K0   A0   B0   C0   D0
K1   A1   B1  NaN  NaN
K2   A2   B2   C2   D2
K3  NaN  NaN   C3   D3

# Operations

There are lots of operations with pandas that will be really useful to you, but don't fall into any distinct category.

In [42]:
df = pd.DataFrame({"col1": [1, 2, 3, 4], "col2": [444, 555, 666, 444], "col3": ["abc", "def", "ghi", "xyz"]})
df

col1  col2 col3
0     1   444  abc
1     2   555  def
2     3   666  ghi
3     4   444  xyz

## Info on Unique Values

In [43]:
df['col2'].unique()

array([444, 555, 666])

In [44]:
df['col2'].nunique()

3

In [48]:
df['col2'].value_counts()

444    2
555    1
666    1
Name: col2, dtype: int64

## Selecting data

In [108]:
new_df = df[(df["col1"] > 2) & (df["col2"] == 444)]

In [109]:
new_df

col1  col2 col3
3     4   444  xyz

## Applying Functions

Really important thing used all the time - especially with lambda.

In [49]:
def times_2(x):
    return x * 2

In [51]:
df['col1'].apply(times_2)

0    2
1    4
2    6
3    8
Name: col1, dtype: int64

In [57]:
df['col1'].apply(lambda x: x * 2)

0    2
1    4
2    6
3    8
Name: col1, dtype: int64

In [112]:
df["col3"].apply(len)

0    3
1    3
2    3
3    3
Name: col3, dtype: int64

In [113]:
df["col1"].sum()

10

## Permanently Removing a Column

In [114]:
df

col1  col2 col3
0     1   444  abc
1     2   555  def
2     3   666  ghi
3     4   444  xyz

In [58]:
df.drop("col1", axis=1)

col2 col3
0   444  abc
1   555  def
2   666  ghi
3   444  xyz

or 

In [115]:
del df["col1"]

In [116]:
df

col2 col3
0   444  abc
1   555  def
2   666  ghi
3   444  xyz

## Get column and index names.

In [59]:
df.columns

Index(['col1', 'col2', 'col3'], dtype='object')

In [118]:
df.index

RangeIndex(start=0, stop=4, step=1)

## Sorting and Ordering a DataFrame

In [52]:
df

col1  col2 col3
0     1   444  abc
1     2   555  def
2     3   666  ghi
3     4   444  xyz

In [53]:
df.sort_values(by="col2") # inplace=False by default

col1  col2 col3
0     1   444  abc
3     4   444  xyz
1     2   555  def
2     3   666  ghi

Indexes are sorted too!

## Find Null Values or Check for Null Values

In [54]:
df.isnull()

col1   col2   col3
0  False  False  False
1  False  False  False
2  False  False  False
3  False  False  False

In [55]:
# Drop rows with NaN Values
df.dropna()

col1  col2 col3
0     1   444  abc
1     2   555  def
2     3   666  ghi
3     4   444  xyz

## Pivot tables (Excel feature)

In [126]:
data = {'A':['foo','foo','foo','bar','bar','bar'],
        'B':['one','one','two','two','one','one'],
        'C':['x','y','x','y','x','y'],
        'D':[1,3,2,5,4,1]}

df = pd.DataFrame(data)

In [127]:
df

A    B  C  D
0  foo  one  x  1
1  foo  one  y  3
2  foo  two  x  2
3  bar  two  y  5
4  bar  one  x  4
5  bar  one  y  1

In [128]:
df.pivot_table(values="D", index=["A", "B"], columns=["C"])

C          x    y
A   B            
bar one  4.0  1.0
    two  NaN  5.0
foo one  1.0  3.0
    two  2.0  NaN

# Data Input and Output

Pandas can read a variety of file types using its pd.read_ methods. Let's take a look at the most common data types.

## CSV

### CSV input

In [63]:
pwd

'/Users/michalsznajder/Dropbox/github/learning_data_science/udemy_python_for_data_science'

In [64]:
df = pd.read_csv('example') # you can hit tab after "exa" for autocompletion!
df

a   b   c   d
0   0   1   2   3
1   4   5   6   7
2   8   9  10  11
3  12  13  14  15

You can see wide variety of data sources that Pandas can read from by entering `pd.read_[tab]`.

In [ ]:
pd.read_ [tab]

### CSV Output

In [130]:
df.to_csv("example_out", index=False) # index=False we do not want to save index here.

Similarly - you can get all methods for writing data by doing `pd.to_[tab]`.

In [ ]:
df.to_ [tab]

## Excel

Pandas can read and write excel files, keep in mind, this only imports data. Not formulas or images. Having images or macros may cause this read_excel method to crash.

### Excel Input

In [131]:
pd.read_excel("Excel_Sample.xlsx", sheetname="Sheet1")

a   b   c   d
0   0   1   2   3
1   4   5   6   7
2   8   9  10  11
3  12  13  14  15

### Excel Output

In [132]:
df.to_excel("Excel_Sample.xlsx", sheet_name="Sheet1")

## HTML

You may need to install htmllib5,lxml, and BeautifulSoup4. In your terminal/command prompt run:

    conda install lxml
    conda install html5lib
    conda install BeautifulSoup4

Then restart Jupyter Notebook.
(or use pip install if you aren't using the Anaconda Distribution)

Pandas can read table tabs off of html.

## HTML Input - web scrapping in Pandas

Pandas read_html function will read tables off of a webpage and return a list of DataFrame objects.

In [65]:
df = pd.read_html('http://www.fdic.gov/bank/individual/failed/banklist.html')

In [68]:
type(df)

list

Pandas converts each HTML table element into DataFrame and store it in a list.

In [70]:
df[0].head()

Bank Name                City  ST  \
0                                Fayette County Bank          Saint Elmo  IL   
1  Guaranty Bank, (d/b/a BestBank in Georgia & Mi...           Milwaukee  WI   
2                                     First NBC Bank         New Orleans  LA   
3                                      Proficio Bank  Cottonwood Heights  UT   
4                      Seaway Bank and Trust Company             Chicago  IL   

    CERT                Acquiring Institution      Closing Date   Updated Date  
0   1802            United Fidelity Bank, fsb      May 26, 2017  July 26, 2017  
1  30003  First-Citizens Bank & Trust Company       May 5, 2017  July 26, 2017  
2  58302                         Whitney Bank    April 28, 2017  July 26, 2017  
3  35495                    Cache Valley Bank     March 3, 2017   May 18, 2017  
4  19328                  State Bank of Texas  January 27, 2017   May 18, 2017

# SQL

Pandas is not the best way to read SQL database. That is because there are many SQL engines and it will probably be better to use driver specific module to read the database.

The pandas.io.sql module provides a collection of query wrappers to both facilitate data retrieval and to reduce dependency on DB-specific API. Database abstraction is provided by SQLAlchemy if installed. In addition you will need a driver library for your database. Examples of such drivers are psycopg2 for PostgreSQL or pymysql for MySQL. For SQLite this is included in Python’s standard library by default. You can find an overview of supported drivers for each SQL dialect in the SQLAlchemy docs.

If SQLAlchemy is not installed, a fallback is only provided for sqlite (and for mysql for backwards compatibility, but this is deprecated and will be removed in a future version). This mode requires a Python database adapter which respect the Python DB-API.

The key functions are:

* read_sql_table(table_name, con[, schema, ...])	
    * Read SQL database table into a DataFrame.
* read_sql_query(sql, con[, index_col, ...])	
    * Read SQL query into a DataFrame.
* read_sql(sql, con[, index_col, ...])	
    * Read SQL query or database table into a DataFrame.
* DataFrame.to_sql(name, con[, flavor, ...])	
    * Write records stored in a DataFrame to a SQL database.
    
First let's create a simple SQL engine in memory.

In [77]:
from sqlalchemy import create_engine

In [78]:
engine = create_engine('sqlite:///:memory:')

We take dataframe df and save it into the database.

In [79]:
df = pd.read_excel("Excel_Sample.xlsx", sheetname="Sheet1")
df

a   b   c   d
0   0   1   2   3
1   4   5   6   7
2   8   9  10  11
3  12  13  14  15

In [80]:
df.to_sql('data', engine)

Read data from database.

In [82]:
sql_df = pd.read_sql('data',con=engine)

In [83]:
sql_df

index   a   b   c   d
0      0   0   1   2   3
1      1   4   5   6   7
2      2   8   9  10  11
3      3  12  13  14  15

Again, it is better to use SQL engine specific module rather than Pandas.